### DISCUSSION TITLE
Is steel scheme a scheme?

### DISCUSSION THREAD

**Discussion by amano-kenji** [mattwparas/steel#145] [Q&A] [answered]
I found

> At the moment, it is not explicitly compliant with any individual scheme specification.

on README.

I thought scheme was based around specification? The latest is R7RS.

Older scheme implementations tend to lag behind the latest specification. I don't know what to think of this fragmentation issue.

Am I supposed to think that steel scheme is inspired by scheme and is actually an independent language?

**Comment by mattwparas** [comment, marked as answer]
When I say it is not explicitly compliant, I do not mean that I'm not aiming for compliance, but rather that 100% compliance has not been achieved.

The goal is to be able to run compliant scheme code in a relatively portable way, by bringing in compatibility libraries - the only major fundamental difference currently is that steel does not have mutable pairs by default - by default they are immutable. But that is something that can be fixed with a compatibility library

**Reply by amano-kenji** [reply]
Does it try to stick to R7RS? What if R8RS has breaking changes?

Will steel scheme have persistent data structures from haskell and clojure? Immutable data structures are simple, but persistent data structures are complex to implement.

**Reply by mattwparas** [reply]
The spec is just a spec. It is a minimum set of functions / behaviors. I can assure you that the semantics of steel will match the semantics of scheme, and I will do my best effort to abide by the specs as much as possible. As of now I am almost compliant with r5rs, and after reaching compliance there will move to getting r7rs. However, much like racket, I reserve the right to branch off and do my own thing, while maintaining compatibility libraries.

Steel already has persistent data structures - Vectors (RBB trees), hash maps (Hash array mapped tries), hash sets, lists

**Reply by amano-kenji** [reply]
If you can branch off, then it's a good idea to write a manual on steel scheme. Right now, there is no manual.

**Reply by mattwparas** [reply]
I am working on docs, you can find them hosted here https://mattwparas.github.io/steel/book/

As you can see there are gaps, but its improving over time
